File: packages/xml-zig/source/character.zig
const std = @import("std");

// [2] Char
// Spec: https://www.w3.org/TR/xml/#NT-Char
pub fn isCharacter(code_point: u21) bool {
    return switch (code_point) {
        '\t', '\n', '\r', ' '...0xD7FF, 0xE000...0xFFFD, 0x10000...0x10FFFF => true,
        else => false,
    };
}

pub fn isDiscouragedCharacter(code_point: u21) bool {
    return switch (code_point) {
        0x7F...0x84, 0x86...0x9F, 0xFDD0...0xFDEF, 0x1FFFE...0x1FFFF, 0x2FFFE...0x2FFFF, 0x3FFFE...0x3FFFF, 0x4FFFE...0x4FFFF, 0x5FFFE...0x5FFFF, 0x6FFFE...0x6FFFF, 0x7FFFE...0x7FFFF, 0x8FFFE...0x8FFFF, 0x9FFFE...0x9FFFF, 0xAFFFE...0xAFFFF, 0xBFFFE...0xBFFFF, 0xCFFFE...0xCFFFF, 0xDFFFE...0xDFFFF, 0xEFFFE...0xEFFFF, 0xFFFFE...0xFFFFF, 0x10FFFE...0x10FFFF => true,
        else => false,
    };
}

// Spec:
pub fn isNameCharacter(code_point: u21) bool {
    return isNameStartCharacter(code_point) or switch (code_point) {
        '-', '.', '0'...'9', 0x00B7, 0x0300...0x036F, 0x203F...0x2040 => true,
        else => false,
    };
}

// Spec:
pub fn isNameStartCharacter(code_point: u21) bool {
    return switch (code_point) {
        ':', '_', 'A'...'Z', 'a'...'z', 0x00C0...0x00D6, 0x00D8...0x00F6, 0x00F8...0x02FF, 0x0370...0x037D, 0x037F...0x1FFF, 0x200C...0x200D, 0x2070...0x218F, 0x2C00...0x2FEF, 0x3001...0xD7FF, 0xF900...0xFDCF, 0xFDF0...0xFFFD, 0x10000...0xEFFFF => true,
        else => false,
    };
}

// Spec: https://www.w3.org/TR/xml/#NT-PubidChar
pub fn isPublicIdentifierCharacter(code_point: u21) bool {
    return switch (code_point) {
        ' ', '\r', '\n', 'a'...'z', 'A'...'Z', '0'...'9', '-', '\'', '(', ')', '+', ',', '.', '/', ':', '=', '?', ';', '!', '*', '#', '@', '$', '_', '%' => true,
        else => false,
    };
}

// Spec: https://www.w3.org/TR/xml/#sec-CharData
pub fn isQuote(code_point: u21) bool {
    return switch (code_point) {
        '\'', '"' => true,
        else => false,
    };
}

// [3] S
// Spec: https://www.w3.org/TR/xml/#NT-S
pub fn isWhitespace(code_point: u21) bool {
    return switch (code_point) {
        '\r', '\t', '\n', ' ' => true,
        else => false,
    };
}

test "isNameCharacter" {
    try std.testing.expect(isNameCharacter('A'));
    try std.testing.expect(isNameCharacter('-'));
    try std.testing.expect(isNameCharacter('.'));
    try std.testing.expect(isNameCharacter('9'));
    try std.testing.expect(isNameCharacter(0xB7));
    try std.testing.expect(!isNameCharacter(' '));
    try std.testing.expect(!isNameCharacter(0x0));
}

test "isNameStartCharacter" {
    try std.testing.expect(isNameStartCharacter(':'));
    try std.testing.expect(isNameStartCharacter('_'));
    try std.testing.expect(isNameStartCharacter('A'));
    try std.testing.expect(isNameStartCharacter('z'));
    try std.testing.expect(isNameStartCharacter(0xC0));
    try std.testing.expect(!isNameStartCharacter('0'));
    try std.testing.expect(!isNameStartCharacter('-'));
    try std.testing.expect(!isNameStartCharacter(0x37E)); // Greek question mark, excluded
}

test "isWhitespace" {
    try std.testing.expect(isWhitespace(' '));
    try std.testing.expect(isWhitespace('\t'));
    try std.testing.expect(isWhitespace('\r'));
    try std.testing.expect(isWhitespace('\n'));
    try std.testing.expect(!isWhitespace('a'));
    try std.testing.expect(!isWhitespace(0x0));
}

File: packages/xml-ts/source/character.ts
import { Ascii } from "./ascii.js";

export function isChar(codePoint: number): boolean {
	return (
		codePoint === Ascii.HorizontalTab ||
		codePoint === Ascii.LineFeed ||
		codePoint === Ascii.CarriageReturn ||
		(codePoint >= Ascii.Space && codePoint <= 0xd7ff) ||
		(codePoint >= 0xe000 && codePoint <= 0xfffd) ||
		(codePoint >= 0x10000 && codePoint <= 0x10ffff)
	);
}

/**
 * Checks whether an entire string is one valid XML 'Name'.
 */
export function isName(str: string): boolean {
	if (str.length === 0) {
		return false;
	}

	// Convert string to an array of characters to properly handle Unicode surrogate pairs.
	// e.g., Characters in the #x10000-#xEFFFF range take up two 16-bit code units.
	const chars = Array.from(str);

	// The first character must strictly be a NameStartChar
	const firstCodePoint = chars[0].codePointAt(0);

	if (firstCodePoint === undefined || !isNameStartChar(firstCodePoint)) {
		return false;
	}

	// All subsequent characters must be NameChars
	for (let i = 1; i < chars.length; i++) {
		const codePoint = chars[i].codePointAt(0);

		if (codePoint === undefined || !isNameChar(codePoint)) {
			return false;
		}
	}

	return true;
}

/**
 * Checks whether an entire string is a valid space-separated list of XML 'Name' values.
 */
export function isNames(str: string): boolean {
	if (str.length === 0) {
		return false;
	}

	// The specification strictly requires space (#x20) as the delimiter,
	// not general whitespace (S).
	const namesList = str.split(" ");

	for (const name of namesList) {
		if (!isName(name)) {
			return false;
		}
	}

	return true;
}

export function isNameChar(codePoint: number): boolean {
	return (
		isNameStartChar(codePoint) ||
		(codePoint >= Ascii.Zero && codePoint <= Ascii.Nine) || // 0-9
		codePoint === Ascii.Minus || // -
		codePoint === Ascii.Period || // .
		codePoint === 0x00b7 ||
		(codePoint >= 0x0300 && codePoint <= 0x036f) ||
		(codePoint >= 0x203f && codePoint <= 0x2040)
	);
}

export function isNameStartChar(codePoint: number): boolean {
	return (
		(codePoint >= Ascii.UppercaseA && codePoint <= Ascii.UppercaseZ) || // A-Z
		(codePoint >= Ascii.LowercaseA && codePoint <= Ascii.LowercaseZ) || // a-z
		codePoint === Ascii.Colon || // :
		codePoint === Ascii.Underscore || // _
		(codePoint >= 0x00c0 && codePoint <= 0x00d6) ||
		(codePoint >= 0x00d8 && codePoint <= 0x00f6) ||
		(codePoint >= 0x00f8 && codePoint <= 0x02ff) ||
		(codePoint >= 0x0370 && codePoint <= 0x037d) ||
		(codePoint >= 0x037f && codePoint <= 0x1fff) ||
		(codePoint >= 0x200c && codePoint <= 0x200d) ||
		(codePoint >= 0x2070 && codePoint <= 0x218f) ||
		(codePoint >= 0x2c00 && codePoint <= 0x2fef) ||
		(codePoint >= 0x3001 && codePoint <= 0xd7ff) ||
		(codePoint >= 0xf900 && codePoint <= 0xfdcf) ||
		(codePoint >= 0xfdf0 && codePoint <= 0xfffd) ||
		(codePoint >= 0x10000 && codePoint <= 0xeffff)
	);
}

export type Quote = typeof Ascii.DoubleQuote | typeof Ascii.SingleQuote;

export function isQuote(codePoint: number): codePoint is Quote {
	return codePoint === Ascii.DoubleQuote || codePoint === Ascii.SingleQuote;
}

export function isRestrictedChar(codePoint: number): boolean {
	return (
		(codePoint >= Ascii.StartOfHeader && codePoint <= Ascii.Backspace) ||
		(codePoint >= Ascii.VerticalTab && codePoint <= Ascii.FormFeed) ||
		(codePoint >= Ascii.ShiftOut && codePoint <= Ascii.UnitSeparator) ||
		(codePoint >= Ascii.Delete && codePoint <= 0x84) ||
		(codePoint >= 0x86 && codePoint <= 0x9f)
	);
}

export function isWhitespace(codePoint: number): boolean {
	return (
		codePoint === Ascii.CarriageReturn ||
		codePoint === Ascii.HorizontalTab ||
		codePoint === Ascii.LineFeed ||
		codePoint === Ascii.Space
	);
}
